gem/ci/build.ps1

```diff
@@ -0,0 +1,197 @@
+$ErrorActionPreference = 'Stop'
+
+function Get-ContainerVersion
+{
+    # Don't need to fetch anything else when running in Azure CI, as that will
+    # already have pulled everything down, and more importantly git-fetch will
+    # block waiting for authorization. TF_BUILD is set by Azure CI.
+    if ($ENV:TF_BUILD -ine 'true')
+    {
+        # shallow repositories need to pull remaining code to `git describe` correctly
+        if (Test-Path "$(git rev-parse --git-dir)/shallow")
+        {
+            git pull --unshallow
+        }
+
+        # tags required for versioning
+        git fetch origin 'refs/tags/*:refs/tags/*'
+    }
+
+    (git describe) -replace '-.*', ''
+}
+
+# only need to specify -Name or -Path when calling
+function Lint-Dockerfile(
+    $Name,
+    $Path = "docker/$Name/Dockerfile",
+    $Ignore = @())
+{
+    $ignores = @()
+    foreach ($code in @('DL3008','DL3018','DL4000','DL4001') + $Ignore) {
+        $ignores += '--ignore', $code
+    }
+
+    Write-Output "hadolint $ignores $Path"
+
+    hadolint $ignores $Path
+}
+
+function Build-Container(
+    $Name,
+    $Namespace = 'puppet',
+    $Dockerfile = "docker/$Name/Dockerfile",
+    $Context = "docker/$Name",
+    $Version = (Get-ContainerVersion),
+    $Vcs_ref = $(git rev-parse HEAD),
+    $Pull = $true)
+{
+    $build_date = (Get-Date).ToUniversalTime().ToString('o')
+    $docker_args = @(
+        '--build-arg', "version=$Version",
+        '--build-arg', "vcs_ref=$Vcs_ref",
+        '--build-arg', "build_date=$build_date",
+        '--build-arg', "namespace=$Namespace",
+        '--file', "$Dockerfile",
+        '--tag', "$Namespace/${Name}:$Version"
+    )
+
+    if ($Pull) {
+        $docker_args += '--pull'
+    }
+
+    Write-Output "docker build $docker_args $Context"
```

Comment:
Looks like we have a lingering `Write-Output` here

Reply:
Hm?

That's intentional.

Reply:
Remember we tried using other `Write-` methods but they didn't work.

I just reverted to the initial code that worked.

Feel free to refactor all these Write- calls to something that is better.